src/composition/supergraph/config/resolve/mod.rs

```diff
@@ -227,8 +259,8 @@ impl FullyResolvedSubgraphs {
     }
 
     /// Used to upsert a fully resolved subgraph into this object's definitions
-    pub fn upsert_subgraph(&mut self, name: String, schema: String) {
-        self.subgraphs.insert(name, schema);
+    pub fn upsert_subgraph(&mut self, name: String, schema: String) -> bool {
+        self.subgraphs.insert(name, schema).is_none()
```

Comment:
Slightly changing the upsert mechanism allows us to see if we're adding a subgraph for the first time so we can emit the right events for the LSP

Reply:
don't _love_ this return type because I'm not sure what it means (outside of this comment); worth calling it out in its comment why it returns a bool? or, how do you feel about a more self-documenting enum whose variants capture what we'd want to know? (or, maybe better: should we just move the logic for telling whether we're adding subgraphs for the first time to the bit of code that actually cares about it? that'd better separate what we're up to)

Reply:
Another option might be to just rename this as `insert_subgraph` and pass back the return type from the `insert` method? That's probably more idiomatic

Reply:
that sounds good!